### PR TITLE
[SYCL] Switch build of SYCL runtime to C++17

### PR DESCRIPTION
**Pull request by vladimirlaz**
- switch build of SYCL RT and SYCL unit tests to use C++17 standard;
 - fix build issues exposed by build with VS 2017 compiler;
 - workaround problem in googletest (google/googletest#1616);
 - update documentation accordingly.

### CODE REVIEW

**Comment by pvchupin** [comment]
In general it's good direction. We should carefully understand impact before merge (and fix associated issues).

**Review by pvchupin — CHANGES_REQUESTED** [review]
Just gating it from accidental merge.

**Comment by v-klochkov** [comment]
Interesting..., 9bc9b258b9d166a08426d734f428968a878d336a could be built on Windows locally.
I used 'python buildbot/configure.py; python buildbot/compile.py'. It looks like CI builds differently.

**Comment by v-klochkov** [comment]
> Or it passes locally because of different version of MSVC. I have this:
> "Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29336 for x64"

@vladimirlaz 
See the example (extract from our sources). https://godbolt.org/z/cc8cqr4ET
It passes with MSVC versions starting from 19.27, and fails with 19.26 and older.

**Comment by pvchupin** [comment]
@vladimirlaz, please update https://github.com/intel/llvm/blob/sycl/sycl/doc/GetStartedGuide.md#c-standard along with this change

**Comment by vladimirlaz** [comment]
> @vladimirlaz, please update https://github.com/intel/llvm/blob/sycl/sycl/doc/GetStartedGuide.md#c-standard along with this change

done

File: sycl/doc/GetStartedGuide.md
Comment (bader) [inline sycl/doc/GetStartedGuide.md:642]: @vladimirlaz, could you confirm that there are no runtime library ABI breaking changes requiring us to increment the library version, please?

Comment (v-klochkov) [reply sycl/doc/GetStartedGuide.md:642]: Idea for a separate PR: It would be good to have a static_assert somewhere in SYCL headers that would check that C++ is 17 or higher.
It would give user-friendly message instead of some unexpected compilation errors due to various reasons (using 'if constexpr', auto, inline vars, etc. with C++14 option)

**Comment by kbsmith-intel** [comment]
I am very concerned about how changing over to C++ 17 requirement might adversely impact users.  Seems like this is an API breaking change, and would require rolling major version of sycl headers and libraries.

**Review by pvchupin — CHANGES_REQUESTED** [review]
Holding off for more testing.

**Review by kbobrovs — COMMENTED** [review]
We should also clean-up code redundant with C++17, like this
https://github.com/intel/llvm/blob/9da31049d74ec15dbf54054ab3a2c45e22a692d4/sycl/include/CL/sycl/detail/stl_type_traits.hpp#L20

Is this going to be a different PR?

**Comment by vladimirlaz** [comment]
> We should also clean-up code redundant with C++17, like this
> https://github.com/intel/llvm/blob/9da31049d74ec15dbf54054ab3a2c45e22a692d4/sycl/include/CL/sycl/detail/stl_type_traits.hpp#L20
> 
> Is this going to be a different PR?

Yes this will be done in separate PR

**Review by v-klochkov — APPROVED** [review]

**Comment by bader** [comment]
@pvchupin, ping.

**Review by pvchupin — APPROVED** [review]

**Comment by pvchupin** [comment]
@vladimirlaz, please add a description to PR what this patch is doing.

**Comment by vladimirlaz** [comment]
> @vladimirlaz, please add a description to PR what this patch is doing.

done

**Pull request merged**
